core/state/statedb.go

```diff
@@ -711,54 +985,127 @@ func (s *StateDB) SystemAddressRedo() bool {
+		}
 	}
 
 	stateObject := s.getStateObject(addr)
 	if stateObject != nil {
+		log.Info("GetCode in main", "txIndex", s.txIndex, "addr", addr)
 		return stateObject.Code(s.db)
 	}
+	log.Info("GetCode not exist", "txIndex", s.txIndex, "addr", addr)
 	return nil
 }
 
 func (s *StateDB) GetCodeSize(addr common.Address) int {
 	if s.parallel.isSlotDB {
 		s.parallel.codeReadsInSlot[addr] = struct{}{} // code size is part of code
+		// 1.Try to get from dirty
+		if obj, ok := s.parallel.dirtiedStateObjectsInSlot[addr]; ok {
+			log.Info("GetCodeSize in dirty", "addr", addr)
+			return obj.CodeSize(s.db)
+		}
+		// 2.Try to get from uncomfirmed
+		if obj := s.getCodeFromUnconfirmedDB(addr); obj != nil {
+			log.Info("GetCodeSize in unconfirmed", "addr", addr)
+			return obj.CodeSize(s.db)
+		}
 	}
 
 	stateObject := s.getStateObject(addr)
 	if stateObject != nil {
+		log.Info("GetCodeSize in main", "addr", addr)
 		return stateObject.CodeSize(s.db)
 	}
+	log.Info("GetCodeSize not exist", "addr", addr)
 	return 0
 }
 
 func (s *StateDB) GetCodeHash(addr common.Address) common.Hash {
 	if s.parallel.isSlotDB {
 		s.parallel.codeReadsInSlot[addr] = struct{}{} // code hash is part of code
+		// 1.Try to get from dirty
+		if obj, ok := s.parallel.dirtiedStateObjectsInSlot[addr]; ok {
+			log.Info("GetCodeHash in dirty", "txIndex", s.txIndex, "addr", addr, "codeHash", obj.CodeHash())
+			return common.BytesToHash(obj.CodeHash())
+		}
+		// 2.Try to get from uncomfirmed
+		if obj := s.getCodeFromUnconfirmedDB(addr); obj != nil {
+			log.Info("GetCodeHash in unconfirmed", "txIndex", s.txIndex, "addr", addr, "codeHash", obj.CodeHash())
+			return common.BytesToHash(obj.CodeHash())
+		}
 	}
 
 	stateObject := s.getStateObject(addr)
 	if stateObject == nil {
+		log.Info("GetCodeHash not exist", "txIndex", s.txIndex, "addr", addr)
 		return common.Hash{}
 	}
+	log.Info("GetCodeHash in main", "txIndex", s.txIndex, "addr", addr, "codeHash", stateObject.CodeHash())
 	return common.BytesToHash(stateObject.CodeHash())
 }
 
 // GetState retrieves a value from the given account's storage trie.
+// For parallel mode wih, get from the state in order:
+//   -> self dirty, both Slot & MainProcessor
+//   -> pending of self: Slot on merge
+//   -> pending of unconfirmed DB
+//   -> pending of main StateDB
+//   -> origin
 func (s *StateDB) GetState(addr common.Address, hash common.Hash) common.Hash {
 	if s.parallel.isSlotDB {
 		if s.parallel.stateReadsInSlot[addr] == nil {
 			s.parallel.stateReadsInSlot[addr] = make(map[common.Hash]struct{}, defaultNumOfSlots)
 		}
 		s.parallel.stateReadsInSlot[addr][hash] = struct{}{}
+		if obj, ok := s.parallel.dirtiedStateObjectsInSlot[addr]; ok {
+			log.Info("GetState addr in dirty", "SlotIndex", s.parallel.SlotIndex, "txIndex", s.txIndex,
+				"addr", addr)
+			// 1.Try to get from dirty
+			if val, ok := obj.dirtyStorage[hash]; ok {
+				log.Info("GetState key in dirty", "SlotIndex", s.parallel.SlotIndex, "txIndex", s.txIndex,
```

Comment:
don't think directly read `dirtyStorage` and `pendingStorage` from StateObject is a good idea in StateDB